libsolidity/ast/Types.h

```diff
@@ -372,6 +373,14 @@ class Type
 	/// are returned without modification.
 	virtual TypeResult interfaceType(bool /*_inLibrary*/) const { return nullptr; }
 
+	/// @returns list of subtypes for composite types (base type for array, members for structs,
+	/// value type for mappings) or empty list for ordinary types
+	virtual std::vector<Type const*> subtypes() const { return {}; }
+	/// @returns list of collision-prone subtypes (with an approximate size exceeding 2**64)
+	std::vector<Type const*> storageOversizedSubtypes() const;
+	/// @returns true if this is valid to be stored in data location _loc
+	BoolResult validForLocation(DataLocation _loc, std::set<std::string>* io_seen = nullptr) const;
```

Comment:
Maybe it would make sense to rename this function now to something like `oversizedForLocation()` (and negate its value in that case). It used to be virtual and every derived type could theoretically add add something other than a size check. But now the size check is hard-coded in the base class.

Actually, I wonder why `MappingType` did not have this function overridden to return `false` for memory and calldata.

Reply:
I think there are two semantic variants:
* `validForLocation` should do what the name says; currently it checks sizes (via virtual functions), but it should, at least conceptually, check other properties (perhaps via other virtual functions);
* `validForLocation` does just type size checks, and not intended for anything else.

In the former case we'd extend the implementation, in the later case we should rename the function, as you suggested.

Reply:
I agree. My suggestion comes from the fact that it's not really doing these other checks even in cases where it could and should. It was probably intended to have the first, broader, interpretation but it ended up with a narrower implementation so changing the name would clarify that.